src/promptflow/promptflow/executor/_async_nodes_scheduler.py

```diff
@@ -117,3 +126,57 @@ async def _sync_function_to_async_task(
         return await asyncio.get_running_loop().run_in_executor(
             executor, context.invoke_tool, node, f, kwargs
         )
+
+
+def signal_handler(sig, frame):
+    """
+    Start a thread to monitor coroutines after receiving signal.
+    """
+    flow_logger.info(f"Received signal {sig}({signal.Signals(sig).name}),"
+                     " start coroutint monitor thread.")
+    loop = asyncio.get_running_loop()
+    monitor = threading.Thread(target=monitor_coroutine_after_cancellation, args=(loop,))
+    monitor.start()
+    raise KeyboardInterrupt
+
+
+def monitor_coroutine_after_cancellation(loop: asyncio.AbstractEventLoop):
+    """Exit the process when all coroutines are done.
+    We add this function because if a sync tool is running in async mode,
+    the task will be cancelled after receiving SIGINT,
+    but the thread will not be terminated and blocks the program from exiting.
+    :param loop: event loop of main thread
+    :type loop: asyncio.AbstractEventLoop
+    """
+    # TODO: Use environment variable to ensure it is flow test / node test scenario
```

Comment:
Nitpick: node test will not use this `scheduler`.
We just use context to invoke the tool.
![image](https://github.com/microsoft/promptflow/assets/17527303/729dfdf6-fc20-4528-8730-ac9ec63c8ec3)

Reply:
Learned. Thanks Robben.